central/cloudsources/manager/manager.go

```diff
@@ -202,18 +222,96 @@ func (m *managerImpl) matchDiscoveredClusters(clusters []*discoveredclusters.Dis
 		return
 	}
 
+	concurrency.WithLock(&m.unspecifiedProviderLock, func() {
+		m.unspecifiedProviderTypes = unspecifiedProviderTypes.Clone()
+	})
+
 	for _, cluster := range clusters {
 		switch {
 		case securedClusters.Contains(cluster.GetID()):
 			cluster.Status = storage.DiscoveredCluster_STATUS_SECURED
-		case unspecifiedProviderType.Contains(cluster.GetProviderType().String()):
+		case unspecifiedProviderTypes.Contains(cluster.GetProviderType().String()):
 			cluster.Status = storage.DiscoveredCluster_STATUS_UNSPECIFIED
 		default:
 			cluster.Status = storage.DiscoveredCluster_STATUS_UNSECURED
 		}
 	}
 }
 
+func (m *managerImpl) changeStatusForDiscoveredClusters(clusterID string, status storage.DiscoveredCluster_Status) {
+	cluster, err := m.getCluster(clusterID)
+	if err != nil {
+		log.Errorw("Failed to get cluster to change status",
+			logging.ClusterID(clusterID), logging.Err(err))
+		return
+	}
+
+	// If the cluster has no metadata, we can short-circuit since it cannot match with any discovered cluster.
+	if cluster.GetStatus().GetProviderMetadata().GetCluster().GetId() == "" {
+		return
+	}
+
+	cloudSources, err := m.cloudSourcesDataStore.ListCloudSources(cloudSourceCtx, search.EmptyQuery())
+	if err != nil {
+		log.Errorw("Failed to list stored cloud sources for changing cluster status",
+			logging.Err(err), logging.ClusterID(clusterID))
+		return
+	}
+
+	discoveredClusters := m.fetchDiscoveredClusters(cluster, cloudSources)
+	// In case we found no discovered clusters, we can short-circuit here.
+	if len(discoveredClusters) == 0 {
+		return
+	}
+
+	unspecifiedProviders := concurrency.WithRLock1(&m.unspecifiedProviderLock, func() set.StringSet {
+		return m.unspecifiedProviderTypes.Clone()
+	})
+
+	for _, discoveredCluster := range discoveredClusters {
+		// If we reset the status to unsecured, we need to make sure to also respect the unspecified status that
+		// the discovered cluster may contain.
+		// We currently set this to unspecified because we do not distinguish between clusters being removed
+		// or the cluster just lost connectivity.
+		if status == storage.DiscoveredCluster_STATUS_UNSECURED &&
```

Comment:
Let's remove the check and rely on eventual consistent background loop to get rid of the lock and shared set.